cmd/sfncli/sfncli.go

```diff
@@ -179,61 +189,52 @@ func main() {
 			log.TraceD("getactivitytask", logger.M{"input": input, "token": token})
 
 			// Create a context for this task. We'll cancel this context on errors.
-			// Anything spawned on behalf of the task should use this context.
 			taskCtx, taskCtxCancel := context.WithCancel(mainCtx)
+			taskRunner := NewTaskRunner(*cmd, sfnapi, token, *workDirectory)
+			go func() {
+				defer taskCtxCancel()
+				if err := taskRunner.Process(taskCtx, []string{}, input); err != nil {
+					log.ErrorD("task-error", logger.M{"error": err.Error()})
+				}
+			}()
 
-			// Begin sending heartbeats
+			// heartbeat loop for this task
 			go func() {
 				if err := taskHeartbeatLoop(taskCtx, sfnapi, token); err != nil {
 					log.ErrorD("heartbeat-error", logger.M{"error": err.Error()})
-					// taskHeartBeatLoop only returns errors when they should be treated as critical
-					// e.g., if the task timed out
-					// shut down the command in these cases
-					taskCtxCancel()
-					return
 				}
-				log.TraceD("heartbeat-end", logger.M{"token": token})
 			}()
 
-			// Run the command. Treat unprocessed args (flag.Args()) as additional args to
-			// send to the command on every invocation of the command
-			taskRunner := NewTaskRunner(*cmd, sfnapi, token, *workDirectory)
-			err = taskRunner.Process(taskCtx, flag.Args(), input)
-			if err != nil {
-				log.ErrorD("task-process-error", logger.M{"error": err.Error()})
-				taskCtxCancel()
-				continue
-			}
-
-			// success!
-			taskCtxCancel()
+			// Wait for task to complete
+			<-taskCtx.Done()
```

Comment:
Why was this code changed so significantly? I think it is still doing the same thing but I am wondering if with this change you are introducing a race condition.